pkg/store/bucket.go

```diff
@@ -1908,7 +1909,11 @@ func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []uint64, refetc
 	r.mtx.Unlock()
 
 	for i, id := range ids {
-		c := b[id-start:]
+		seriesOffset := id - start
+		if int(seriesOffset) > indexLength {
```

Comment:
This just checks the offset, but the following code could overread anyway. Am I missing anything?

Reply:
Hi, I found the following issue, we encountered the same issue.
https://github.com/thanos-io/thanos/issues/3057

When I debugged, I found that when using readIndexRange and readChunkRange, the length of the data is not necessarily equal to or longer than the required length of the data, but when we use it, we will set the offset according to the length, so there will be situations .

I use the s3 interface and the underlying ceph storage. I am still looking for ways to solve the root cause of the incorrect data length, but there is no progress. In addition, I hope the store will not panic in this case

I tried to modify the readChunkRange and readIndexRange methods, but I found that other upstream methods do not require the returned data length. So I only modified the place where panic occurs : )

Reply:
@MegaEle do you use radosgw, perchance? I have seen the same thing a few times but never got to it. It's probably some kind of bug there in the API in that it doesn't follow the spec. Have you ever found a way to consistently reproduce it?

Reply:
Not sure if the issue I just encountered is also relevant: https://github.com/thanos-io/thanos/issues/3339

I use Radosgw/ceph to confirm what @GiedriusS says